utoipa-gen/src/lib.rs

```diff
@@ -1245,45 +1238,111 @@ struct ArrayOrOptionType<'a>(Type<'a>);
 
 impl Parse for ArrayOrOptionType<'_> {
     fn parse(input: ParseStream) -> syn::Result<Self> {
-        let mut is_array = false;
-        let mut is_option = false;
-
-        let mut parse_array = |input: &ParseBuffer| {
-            is_array = true;
-            let group;
-            bracketed!(group in input);
-            group.parse::<Ident>()
-        };
-
-        let ty = if input.peek(syn::Ident) {
-            let mut ident: Ident = input.parse()?;
-
-            // is option of type or [type]
-            if (ident == "Option" && input.peek(Token![<]))
-                && (input.peek2(syn::Ident) || input.peek2(Bracket))
-            {
-                is_option = true;
+        const EXPECTED_TYPE_MESSAGE: &str =
+            "Expected a type/path such as path::to::Foo, or Foo. May also be Option<Foo> or [Foo].";
+
+        fn parse_type<'a>(t: syn::Type) -> syn::Result<Type<'a>> {
+            let mut is_option: bool = false;
+            let mut is_array: bool = false;
+            let path: TypePath = match t {
+                syn::Type::Path(mut path) => {
+                    let type_segment: &PathSegment =
+                        path.path.segments.last().ok_or_else(|| {
+                            syn::Error::new(path.path.span(), "No last path segment")
+                        })?;
+                    let ident = &type_segment.ident;
+
+                    // is option of type or [type]
+                    if ident == "Option" {
+                        is_option = true;
+
+                        let angle_bracketed: &AngleBracketedGenericArguments = match &type_segment
+                            .arguments
+                        {
+                            PathArguments::AngleBracketed(angle_bracketed) => angle_bracketed,
+                            _ => {
+                                return Err(syn::Error::new(type_segment.span(), "Option must have its generic type parameter specified. e.g. Option<String>"));
+                            }
+                        };
+
+                        if angle_bracketed.args.len() != 1 {
+                            return Err(syn::Error::new(type_segment.span(), "Option must have only a single generic parameter specified. e.g. Option<String>"));
+                        }
+
+                        let argument: &GenericArgument = angle_bracketed.args.first().expect(
+                            "Expected there to be 1 angle bracketed argument for Option<...>",
+                        );
+
+                        let argument_path: &TypePath = match argument {
+                            GenericArgument::Type(syn::Type::Path(path)) => path,
```

Comment:
We probably could get rid of this whole custom type parsing and use the `ComponentPart`. This would allow users to use the regular Rust types as arguments.

Reply:
Unfortunately it fails to parse things like `[Foo]` and `Option<[Foo]>`

Reply:
Perhaps we could do a breaking change, and make it so people need to use `Vec<Foo>` and `Option<Vec<Foo>>` instead?